apollo-router/src/spec/query/change.rs

```diff
@@ -194,97 +248,127 @@ impl<'a> QueryHashVisitor<'a> {
+        "^TYPE".hash(self);
+
         match t {
-            schema::Type::Named(name) => self.hash_type_by_name(name.as_str()),
+            schema::Type::Named(name) => self.hash_type_by_name(name.as_str())?,
             schema::Type::NonNullNamed(name) => {
                 "!".hash(self);
-                self.hash_type_by_name(name.as_str())
+                self.hash_type_by_name(name.as_str())?;
             }
             schema::Type::List(t) => {
                 "[]".hash(self);
-                self.hash_type(t)
+                self.hash_type(t)?;
             }
             schema::Type::NonNullList(t) => {
                 "[]!".hash(self);
-                self.hash_type(t)
+                self.hash_type(t)?;
             }
         }
+        "^TYPE-END".hash(self);
+        Ok(())
     }
 
     fn hash_field(
         &mut self,
         parent_type: String,
-        type_name: String,
         field_def: &FieldDefinition,
-        arguments: &[Node<Argument>],
+        node: &executable::Field,
     ) -> Result<(), BoxError> {
-        if self.hashed_fields.insert((parent_type.clone(), type_name)) {
-            self.hash_type_by_name(&parent_type)?;
+        "^FIELD".hash(self);
 
-            field_def.name.hash(self);
+        self.hash_type_by_name(&parent_type)?;
 
-            for argument in &field_def.arguments {
-                self.hash_input_value_definition(argument)?;
-            }
+        field_def.name.hash(self);
+        self.hash_type(&field_def.ty)?;
 
-            for argument in arguments {
-                self.hash_argument(argument);
-            }
+        for argument in &field_def.arguments {
+            self.hash_input_value_definition(argument)?;
+        }
+
+        for argument in &node.arguments {
+            self.hash_argument(argument);
+        }
 
-            self.hash_type(&field_def.ty)?;
+        self.hash_type(&field_def.ty)?;
 
-            for directive in &field_def.directives {
-                self.hash_directive(directive);
-            }
+        for directive in &field_def.directives {
+            self.hash_directive(directive);
+        }
+
+        self.hash_join_field(&parent_type, &field_def.directives)?;
 
-            self.hash_join_field(&parent_type, &field_def.directives)?;
+        for directive in &node.directives {
+            self.hash_directive(directive);
         }
+
+        node.alias.hash(self);
+        "^FIELD-END".hash(self);
+
         Ok(())
     }
 
     fn hash_input_value_definition(
         &mut self,
         t: &Node<ast::InputValueDefinition>,
     ) -> Result<(), BoxError> {
+        "^INPUT_VALUE".hash(self);
+
         self.hash_type(&t.ty)?;
         for directive in &t.directives {
             self.hash_directive(directive);
         }
         if let Some(value) = t.default_value.as_ref() {
             self.hash_value(value);
+        } else {
+            "no default value".hash(self);
```

Comment:
Since this edge case is handled in two places maybe it better to change `hash_value` to accept `Option`?

Reply:
Also `no default value` contrasts with other literals used by hashing code since it doesn't start with `^` and use spaces

Reply:
updated in 935ecf91e